server/src/modules/maps/maps.service.ts

```diff
@@ -406,7 +414,160 @@ export class MapsService {
+        if (!uploadedImages) throw new BadGatewayException('Failed to upload image to cdn');
+    }
+
+    async deleteImage(userID: number, imgID: number): Promise<void> {
+        const img = await this.mapRepo.getImages({ id: imgID })
+        if (img.length === 0) throw new NotFoundException('Image not found');
+
+        const map = await this.mapRepo.get(img[0].mapID);
+        if (map.submitterID !== userID) throw new ForbiddenException('User is not the submitter of the map');
+
+        if (map.statusFlag !== MapStatus.NEEDS_REVISION) throw new ForbiddenException('Map is not in NEEDS_REVISION state');
+
+        await this.deleteStoredMapImage(imgID);
+        await this.mapRepo.deleteImage({ id: imgID });
+    }
+
+    async editSaveMapImageFile(
+        imgBuffer: Buffer,
+        fileName: string,
+        width: number,
+        height: number
+    ): Promise<FileStoreCloudFile> {
+        let edited;
+        try {
+            edited = await sharp(imgBuffer)
+                .resize(width, height, { fit: 'inside' })
+                .jpeg({ mozjpeg: true })
+                .toBuffer();
+        } catch { // This looks bad, but sharp is very non-specific about its errors
+            throw new BadRequestException('Invalid image file');
+        }
+        return this.fileCloudService.storeFileCloud(edited, fileName);
+    }
+
+    async storeMapImage(imgBuffer: Buffer, imgID: number): Promise<FileStoreCloudFile[]> {
+        return Promise.all([
+            this.editSaveMapImageFile(imgBuffer, `img/${imgID}-small.jpg`, 480, 360),
+            this.editSaveMapImageFile(imgBuffer, `img/${imgID}-medium.jpg`, 1280, 720),
+            this.editSaveMapImageFile(imgBuffer, `img/${imgID}-large.jpg`, 1920, 1080)
+        ]);
+    }
+
+    async deleteStoredMapImage(imgID: number): Promise<void> {
+        await Promise.all([
+            this.fileCloudService.deleteFileCloud(`img/${imgID}-small.jpg`),
+            this.fileCloudService.deleteFileCloud(`img/${imgID}-medium.jpg`),
+            this.fileCloudService.deleteFileCloud(`img/${imgID}-large.jpg`)
+        ]);
+    }
+    //#endregion
+
+    //#region Map Thumbnail
+
+    async updateThumbnail(userID: number, mapID: number, imgBuffer: Buffer): Promise<void> {
+        let map = await this.mapRepo.get(mapID, { thumbnail: true });
+
+        if (!map) throw new NotFoundException('Map not found');
+        if (map.submitterID !== userID) throw new ForbiddenException('User is not the submitter of the map');
+        if (map.statusFlag !== MapStatus.NEEDS_REVISION) throw new ForbiddenException('Map is not in NEEDS_REVISION state');
+
+        if (!map.thumbnailID) {
+            const newThumbnail = await this.mapRepo.createImage(mapID);
+            map = await this.mapRepo.update(mapID, { thumbnail: { connect: { id: newThumbnail.id } }});
+        }
+
+        const thumbnails = await this.mapRepo.getImages({ id: map.thumbnailID });
+        const thumbnail = thumbnails[0]; // why.
```

Comment:
Funny variable name, why not do `const images = ...`, `const thumbnail = images[0]`

Reply:
Oh see what's happening - you should define a `getImage` method that does a `findUnique` - `id` is obvs going to be a valid WhereUniqueInput.